### ISSUE TITLE
Enable setting flexible gridded reader (and input args) in Colocator 

### ISSUE THREAD

**Issue by jgliss** [metno/pyaerocom#113]
So far, reading of gridded data (e.g. model or satellite level 3 data) was supported only in AeroCom conventions through `ReadGridded`. Now, with other reading interfaces becoming available (e.g. ReadEMEP, cf. #111), this needs to be supported in the highlevel colocation routines.

While on the longer term it would be desirable to have a factory gridded reading object (similar to `ReadUngridded`) that figures out which gridded reader to use, for now it will be good to enable to set the gridded reader in the `Colocator` class. 

Idea is to add an attribute `gridded_reader_id` to `ColocationSetup` together with a dictionary in the header of `Colocator` class that maps supported reader IDs with readers and a dedicated method `get_gridded_reader` that returns the correct reader, like:

```python
class ColocationSetup(BrowseDict):
   
   def __init__(*args, **kwargs):
      self.gridded_reader_id = {
         'model' : 'ReadGridded',
         'obs' : 'ReadGridded'}
      ... # all other attributes

class Colocator(ColocationSetup):
   SUPPORTED_GRIDDED_READERS = {

      'ReadGridded' : pya.io.ReadGridded, 
      'ReadEMEP' : pya.io.ReadEMEP
   }
   def get_gridded_reader(self, what='model'):
      try:
         return self.SUPPORTED_GRIDDED_READERS[self.gridded_reader_id[what]]
      except KeyError:
         # raise proper exception
```

**Comment by jgliss** [comment]
@ejgal I made an issue for what we just discussed, let me know if you have any comments / suggestions.

**Comment by ejgal** [comment]
I implemented a very hackish way of selecting the reader to check if the ReadEMEP class needs more functionality before running the evaluation directly, see [Test changing model reader in colocator](https://github.com/ejgal/pyaerocom/commit/72142fc3932edef068c310cfeecf7191319f0b8b?w=1).

I did an evaluation run with the Aasetal dataset that worked fine, but running against ghost daily with the following code was estimated to take two hours, so I will set up pyaerocom on lustre and run it there.
```
filepath = '/lustre/storeB/project/fou/kl/emep/ModelRuns/2019_REPORTING/EMEP01_L20EC_rv4_33.2018/Base_day.nc'

cfg_dir = 'config'
stp = AerocomEvaluation(proj_id='test', exp_name='Test emep ghost', exp_id='web_emep_ghost', config_dir=cfg_dir)

model_config = {'emep': dict(model_id='EMEP',
                             model_ts_type_read='daily',
                             mtype='?',
                             filepath=filepath,
                             data_id='EMEP',
                             gridded_reader='ReadEMEP')}

obs_config = {'GHOST-daily'        :   dict(obs_id='GHOST.EEA.daily',
                             obs_vars=['concpm10'],
                             obs_vert_type='Surface')}

stp.obs_config = obs_config
stp.model_config = model_config

colset=stp.colocation_settings
colset['filter_name'] = 'WORLD-wMOUNTAINS'
colset['start'] = 2018
colset['colocate_time'] = True
colset['model_ts_type_read'] = 'daily'
colset['ts_type'] = 'daily'
stp.run_evaluation()
```

**Comment by jgliss** [comment]
@ejgal sounds great!

**Comment by ejgal** [comment]
It ran in a much more timely manner on xvis-m4a :)

**Comment by ejgal** [comment]
Updated the code in [emep_web](https://github.com/ejgal/pyaerocom/tree/emep_web) to match better with the suggested code above.

One complication is that ReadEMEP needs additional arguments when instantiating the reader.
I tried solving this by creating a method that instantiates the reader, but not sure if this is the best way to solve it:

```
def instantiate_model_reader(self):
        reader_class = self.get_gridded_reader()
        reader = reader_class(data_id=self.model_id)
        if hasattr(reader, 'filepath'):
            reader.filepath = self.filepath
        return reader
```

**Comment by jgliss** [comment]
> It ran in a much more timely manner on xvis-m4a :)

That is to be expected, bottleneck is likely extraction of timeseries data at site locations and there is likely not much performance to gain. But @hansbrenna certainly has more experience with tweaking the performance.

**Comment by jgliss** [comment]
> Updated the code in [emep_web](https://github.com/ejgal/pyaerocom/tree/emep_web) to match better with the suggested code above.
> 
> One complication is that ReadEMEP needs additional arguments when instantiating the reader.
> I tried solving this by creating a method that instantiates the reader, but not sure if this is the best way to solve it:
> 
> ```
> def instantiate_model_reader(self):
>         reader_class = self.get_gridded_reader()
>         reader = reader_class(data_id=self.model_id)
>         if hasattr(reader, 'filepath'):
>             reader.filepath = self.filepath
>         return reader
> ```

If that works it is fine I think for now. You may be more explicit and do something like 

```python
def instantiate_model_reader(self):
        reader_class = self.get_gridded_reader()
        reader = reader_class(data_id=self.model_id)
        if isinstance(reader_class, ReadEmep) and isinstance(self.filepath, str):
            reader.filepath = self.filepath
        return reader
```

**Comment by jgliss** [comment]
I consider this currently only a temporary solution anyways and some day in the future I imagine reading of GriddedData to be handled via a factory class, similar to `ReadUngridded`, with the gridded readers following a certain template (similar to `ReadUngriddedBase`).

And then, even farther in the future I could imagine to provide an even more highlevel function `read` for reading of any data, e.g.:

```python
import pyaerocom as pya
obsdata = pya.io.read(data_id='AeronetSunV3Lev2.daily', 
                      vars_to_retrieve=['od550aer', 'ang4487aer'])
print(type(obsdata))
# pyaerocom.UngriddedData
modeldata = pya.io.read(data_id='ECMWF_CAMS_REAN', 
                      vars_to_retrieve=['od550aer', 'ang4487aer'])
print(type(modeldata))
# pyaerocom.GriddedData
```

**Comment by ejgal** [comment]
The future plans sounds great! It will be nice to have a common interface for the gridded readers.

Should I also look into adding the possibility of providing a custom gridded_reader_id for observations, or is this not a priority now?

**Comment by jgliss** [comment]
No, this is not a super high priority for now. 

Observations actually already have custom IDs such as `GHOST.EEA.daily`, or `AeronetSunV3Lev2.daily`. And since these observations are *ungridded* in their nature, they are read in as `UngriddedData`. So these should actually not be supported through the *gridded* reading interfaces. 

We should maybe move this into 2 separate issues, 

1. one issue, that addresses the implementation of a factory interface for gridded reading, similar to ungridded reading an
2. one issue that requires 1. to be implemented and addresses a highlevel method `read` that knows about both gridded and ungridded reading interfaces and returns either `UngriddedData` or `GriddedData`.

Feel free to create some issues on that, but I think this should not go into v0.9.0, but into a future release. For v0.9.0 it is more important to add tests and make sure docstrings are provided and up-to-date (and helpful).

**Comment by ejgal** [comment]
Closed by #129
